src/doc/src/reference/semver.md

```diff
@@ -1164,6 +1165,61 @@ Mitigation strategies:
 * Document the platforms and environments you specifically support.
 * Test your code on a wide range of environments in CI.
 
+<a id="new-lints"></a>
+### Minor: introducing new lints
+
+Some changes to a library may cause new lints to be triggered in users of that library.
+This should generally be considered a compatible change.
+
+```rust,ignore,dont-deny
+// MINOR CHANGE
+
+///////////////////////////////////////////////////////////
+// Before
+pub fn foo() {}
+
+///////////////////////////////////////////////////////////
+// After
+#[deprecated]
+pub fn foo() {}
+
+///////////////////////////////////////////////////////////
+// Example use of the library that will safely work.
+
+fn main() {
+    updated_crate::foo(); // Warning: use of deprecated function
+}
+```
+
+Beware that it may be possible for this to technically cause a project to fail if they have explicitly denied the warning, and the updated crate is a direct dependency.
+Denying warnings should be done with care and the understanding that new lints may be introduced over time.
+However, library authors should be cautious about introducing new warnings and may want to consider the potential impact on their users.
+
+The following lints are examples of those that may be introduced when updating a dependency:
+
+* [`deprecated`][deprecated-lint] — Introduced when a dependency adds the [`#[deprecated]` attribute][deprecated] to an item you are using.
+* [`unused_must_use`] — Introduced when a dependency adds the [`#[must_use]` attribute][must-use-attr] to an item where you are not consuming the result.
+* [`unused_unsafe`] — Introduced when a dependency *removes* the `unsafe` qualifier from a function, and that is the only unsafe function called in an unsafe block.
+
+Additionally, updating `rustc` to a new version may introduce new lints.
+
+Transitive dependencies which introduce new lints should not usually cause a failure because Cargo uses [`--cap-lints`](../../rustc/lints/levels.html#capping-lints) to suppress all lints in dependencies.
+
+Mitigating strategies:
+* Options for dealing with denying warnings:
+    * Understand you may need to deal with resolving new warnings whenever you update your dependencies.
+    * Place `deny(warnings)` behind a [feature][Cargo features], for example `#![cfg_attr(feature = "deny-warnings", deny(warnings))]`.
+      Set up your automated CI to check your crate with the feature enabled, possibly as an allowed failure with a notification.
+* Introduce deprecations behind a [feature][Cargo features].
+  For example `#[cfg_attr(feature = "deprecated", deprecated="use bar instead")]`.
+  Then, when you plan to remove an item in a future SemVer breaking change, you can communicate with your users that they should enable the `deprecated` feature *before* updating to remove the use of the deprecated items.
```

Comment:
Unsure if we want to include this but an added bonus is that callers get to decide when to respond to deprecations without a lot of day-to-day noise.

The downside to this being a lot of people don't see/find/follow instructions to enable this.

(both sides are speaking from experience of having done this in clap)

Reply:
I added some text to mention the benefits and drawbacks.

And yea, I imagine communicating to users to use that feature is extremely difficult. I think this is an area where there is opportunity for better options. I keep thinking it is so very similar to `cargo fix --edition` for migrating editions, it would be nice if there was something similar for migrating to new major versions of a dependency. Or at least have some form of better communication.

Reply:
I know someone has worked on a [cargo up](https://crates.io/crates/cargo-up) but the problem with that is its using rust-analyzer for the plugins which the build times ruin it.

Probably the minimal solution is to have some way to include upgrade messages (whether changelogs, migration notes, or links to them).

Probably the next step up is for a new `#[deprecated]` field that is programmatically processable for simple migrations which could only handle the most minimal of deprecations

Reply:
> need to take these manual steps to prepare for a major update.

Hmm, its easy to lose track of the "before" part when talking about deprecations (since its easy to lump in with other breaking changes).  Just having `cargo upgrade` help with these is too late.  `cargo upgrade` could error if there is a message, requiring a flag to explicitly acknowledge that you've done it but people will likely just blindly pass the flag in.

Maybe `cargo upgrade` could do "pre-verify" step and run `cargo check -Awarnings -Ddeprecated` and error if deprecations are present.  Then the problem becomes "what feature flags to use".  `cargo publish` takes in feature flags for its `--verify`, so I guess doing the same.